only replace url for private pipenv source

m3h opened the pull request:
Instead of filtering out private pipenv sources completely, and inserting the urls from the Dependabot credentials configuration (which removes other parts of the pipenv source configuration, e.g. the source name), only replace the url with the new authed url.

This should fix https://github.com/dependabot/dependabot-core/issues/6157

deivid-rodriguez reviewed (commented):
Nice!

This makes sense to me, and the implementation looks good, you just need a few style fixes.

It would be nice though to add a spec to cover this fix, do you think you could do that?

m3h commented:
Cool, I've given it a go - let me know if that works for you.

deivid-rodriguez commented:
Nice, thank you! There's a spec that broke and looks related to the changes in here, can you have a look?

m3h commented:
I've made some changes, but I'm not 100% I've handled it correctly.
I've now 
* added in the auth from the credentials to existing Pipfile sources if the repository URLs can be matched,
* removed existing Pipfile sources that require auth but don't have a match in the credentials
* added all the sources from the credentials.

I have a suspicion that the old approach is based on pipenv scanning _all_ repository sources, which doesn't happen in newer versions of pipenv (see the note [here](https://pipenv.pypa.io/en/latest/advanced/#specifying-package-indexes)).

But I don't really have the historical background with pipenv to comment more on that.

deivid-rodriguez commented:
Interesting, so what you mean is that your initial work that was making one spec fail may be correct in newer piping versions, and we may want to instead change the spec to the new "correct behavior"?

In any case, this PR looks good to me, pending a closer review.

jeffwidman commented:
@m3h can you comment on ☝️ ?

m3h commented:
> Interesting, so what you mean is that your initial work that was making one spec fail may be correct in newer piping versions, and we may want to instead change the spec to the new "correct behavior"?
> 
> In any case, this PR looks good to me, pending a closer review.

No, sorry, I don't think I was being clear. I do think the code needs updating due to the new `pipenv` behaviour. See my comments in https://github.com/dependabot/dependabot-core/issues/6157

The current dependabot behaviour is to replace this:
```
[[source]]
url = "https://${TEST_USER}:${TEST_PASS}@test.pypi.org/simple"
verify_ssl = true
name = "testindex"
```
with this:
```
[[source]]
url = "https://my_test_user:my_test_password@test.pypi.org/simple"
```

stripping the `name` and `verify_ssl` parameters in the process. I was just commenting that this probably used to work when `pipenv` searched *all* sources for a matching package, whereas current versions of `pipenv` will only search sources with a *matching* `name` parameter (which is being stripped by the current dependabot code).


The intent of the PR is to have dependabot instead replace this:
```
[[source]]
url = "https://${TEST_USER}:${TEST_PASS}@test.pypi.org/simple"
verify_ssl = true
name = "testindex"
```
with this:
```
[[source]]
url = "https://my_test_user:my_test_password@test.pypi.org/simple"
verify_ssl = true
name = "testindex"
```

keeping the `verify_ssl` and `name` (and other) parameters intact.

deivid-rodriguez commented on python/lib/dependabot/python/file_updater/pipfile_preparer.rb:
```diff
@@ -114,6 +114,24 @@ def pipfile_sources
+
+            source["url"] = AuthedUrlBuilder.authed_url(credential: source_cred)
+            return source
+
```
The above two lines (`return source` + blank line) are redundant.

m3h replied:
Thanks, I've removed them!

deivid-rodriguez reviewed (commented):
Looks good to me!

I made one minor comment, and I would squash commits into one (I think the second one was just amending the first commit because it broke some specs).

m3h commented:
@deivid-rodriguez  thanks for the comments! I think I've applied them all now.

jeffwidman reviewed (approved):
Thanks for fixing this! Nice job on the debugging and code spelunking to figure out a clean solution.

jeffwidman commented:
I'm going to set auto-merge as I assume CI will pass and looks low risk to merge straight to `main` w/o a deploy, so this will probably get picked up as part of a deploy tomorrow.
